### PR TITLE
 asset based atoms generation

### PR DESCRIPTION
**Pull request by soraphis**
which scripts to generate can now be added to assets.

I've just added it ontop of the current system, so there are some redundancies.

also it has to be tested if this would regenerate assets in packages.

it also contains an custom inspector for Folders, if a folder contains "Atom" it shows a button to generate a new generation file.
not sure if we should keep this, but it could be a nice user workflow improvement. e.g. for typical used atoms (IntVariable / FloatVariable)

### CODE REVIEW

**Comment by bguyl** [comment]
It could be nice to be able to specify an Editor folder
We use asmdef and we keep all editor related code in a single folder, currently I manually drag the generated editor files in the folder

**Comment by miikalo** [comment]
Finally got around checking this out. 

Just as a test, I wanted to create a generator for https://docs.unity3d.com/ScriptReference/Rect.html but the type finder can't find it, these were the only three types found from UnityEngine:

image

If I type it in manually, the generator refuses to generate since it can't find the type:

image
image

This is on a fresh project, I have restarted Unity meanwhile but the problem keeps appearing. I added the Core and BaseAtoms packages only.

Also I can't seem to find this button that should appear if a directory name has "Atom" in it, where is it located?

**Comment by soraphis** [comment]
@miikalo thanks for checking it out!

I fixed the Rect issue, the button should appear when selecting a folder in the inspector.

image

**Comment by miikalo** [comment]
Functionality-wise it's great, however I'd like all of the Atom types to generate be checked by default when creating a generator, I'd more often want to have all of them than none of them and it saves a lot of clicks. Is it also possible to have the editor jump to the newly created generator after pushing the "Add Generator" button?

I have some nits about whitespace and commented out code in the files, after those are addressed it looks good.

File: Packages/Core/Editor/Editors/AtomFolderEditor.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+                var newPath = Path.Combine(
+                    AssetDatabase.GetAssetPath(target),
+                    "New Atom Type.asset"
+                );
+                Debug.Log(newPath);
+                AssetDatabase.CreateAsset(asset, newPath);
+                AssetDatabase.ImportAsset(newPath);
+            }
+        }
+
```
Comment (miikalo) [inline Packages/Core/Editor/Editors/AtomFolderEditor.cs:32]: Nit: extra whitespace

File: Packages/Core/Editor/Editors/AtomFolderEditor.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    public class AtomFolderEditor : UnityEditor.Editor
+    {
+        protected override void OnHeaderGUI()
+        {
+            base.OnHeaderGUI();
+
+            if(! Directory.Exists(AssetDatabase.GetAssetPath(target))) return;
+            if (!target.name.ToLowerInvariant().Contains("atom")) return;
+
+
```
Comment (miikalo) [inline Packages/Core/Editor/Editors/AtomFolderEditor.cs:17]: Nit: extra whitespace

File: Packages/Core/Editor/Editors/AtomGeneratorEditor.cs [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+                    serializedObject.FindProperty("BaseType").stringValue  = s.Split(':')[1];
+                    var type = types.First(grouping => grouping.Key == serializedObject.FindProperty("Namespace").stringValue)
+                        .First(t => t.Name == serializedObject.FindProperty("BaseType").stringValue);
+                    serializedObject.FindProperty("FullQualifiedName").stringValue = type.AssemblyQualifiedName;
+                });
+                dropdown.Show(rect);
+            }
+
+            // EditorGUILayout.PropertyField(serializedObject.FindProperty("Namespace"));
+            // EditorGUILayout.PropertyField(serializedObject.FindProperty("BaseType"));
```
Comment (miikalo) [inline Packages/Core/Editor/Editors/AtomGeneratorEditor.cs:53]: Nit: Can be removed?

File: Packages/Core/Editor/Editors/AtomFolderEditor.cs (hunk 1 above)
Comment (miikalo) [inline Packages/Core/Editor/Editors/AtomFolderEditor.cs:32]: Nit: extra whitespace

File: Packages/Core/Editor/Editors/AtomGeneratorEditor.cs [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+                    serializedObject.FindProperty("Namespace").stringValue = s.Split(':')[0];
+                    serializedObject.FindProperty("BaseType").stringValue  = s.Split(':')[1];
+                    var type = types.First(grouping => grouping.Key == serializedObject.FindProperty("Namespace").stringValue)
+                        .First(t => t.Name == serializedObject.FindProperty("BaseType").stringValue);
+                    serializedObject.FindProperty("FullQualifiedName").stringValue = type.AssemblyQualifiedName;
+                });
+                dropdown.Show(rect);
+            }
+
+            // EditorGUILayout.PropertyField(serializedObject.FindProperty("Namespace"));
```
Comment (miikalo) [inline Packages/Core/Editor/Editors/AtomGeneratorEditor.cs:52]: Nit: commented out code can be removed in this file?

File: Packages/Core/Editor/Editors/AtomGeneratorEditor.cs [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+                    foreach (var type in group)
+                    {
+                        groupItem.AddChild(new AdvancedDropdownItem(group.Key + ":"+ type.Name));
+                    }
+                    root.AddChild(groupItem);
+                }
+                return root;
+            }
+        }
+
```
Comment (miikalo) [inline Packages/Core/Editor/Editors/AtomGeneratorEditor.cs:143]: Nit: extra whitespace

File: Packages/Core/Editor/Generator/AtomGenerator.cs [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+                {
+                    Scripts.Add(null);
+                }
+                idx++;
+                t >>= 1;
+            }
+
+        }
+    }
+
```
Comment (miikalo) [inline Packages/Core/Editor/Generator/AtomGenerator.cs:72]: Nit: extra whitespace

File: Packages/Core/Editor/Generator/RegenerateAllAtoms.cs [hunk 7]
```diff
@@ -33,6 +34,27 @@ public RegenerateItem(string valueType, string baseWritePath, bool isValueEquata
...
+
+            //StringBuilder sb = new StringBuilder();
+            foreach (var guid in guids)
+            {
+                // sb.Append("regenerate ");
+                var path = AssetDatabase.GUIDToAssetPath(guid);
+                // sb.AppendLine(path);
+                AssetDatabase.LoadAssetAtPath<AtomGenerator>(path).Generate();
+            }
+            //Debug.Log(sb.ToString());
```
Comment (miikalo) [inline Packages/Core/Editor/Generator/RegenerateAllAtoms.cs:54]: Nit: commented out code and extra whitespace can be removed?

**Review by miikalo — CHANGES_REQUESTED** [review]

**Review by miikalo — APPROVED** [review]
After a reconsideration, the changes are not absolutely necessary.

**Review by miikalo — APPROVED** [review]

**Pull request merged**